src/flist/db.rs

```diff
@@ -0,0 +1,135 @@
+use anyhow::{bail, Result};
+use futures::StreamExt;
+use md5::{Digest, Md5};
+use std::path::{Path, PathBuf};
+use tokio::fs;
+use tokio::io::{self, AsyncWriteExt};
+use uuid::Uuid;
+pub struct MetadataDbMgr {
+    // root directory where all
+    // the working file of the module will be located
+
+    // underneath are the path for each
+    // sub folder used by the flist module
+    flist: PathBuf,
+}
+
+impl MetadataDbMgr {
+    pub async fn new<P: AsRef<Path>>(flist: P) -> Result<Self> {
+        Ok(Self {
+            flist: flist.as_ref().into(),
+        })
+    }
+
+    pub async fn get<T: AsRef<str>>(&self, url: T) -> Result<PathBuf> {
+        let url = url.as_ref();
+        let hash = self.hash_of_flist(url).await?;
+        let path = self.flist.join(&hash);
+        match fs::File::open(&path).await {
+            Ok(_) => {
+                //Flist already exists let's check it's md5
+                if self.compare_md5(&hash, &path) {
+                    Ok(path)
+                } else {
+                    self.download_flist(url, &hash).await
+                }
+            }
+            Err(error) if error.kind() == io::ErrorKind::NotFound => {
+                self.download_flist(url, &hash).await
+            }
+
+            Err(error) => bail!(
+                "error reading flist file: {}, error {}",
+                &path.display(),
+                error
+            ),
+        }
+    }
+
+    fn compare_md5<T: AsRef<str>, P: AsRef<Path>>(&self, hash: T, path: P) -> bool {
+        // create a Md5 hasher instance
+        let calculated_hash =
+            checksums::hash_file(path.as_ref(), checksums::Algorithm::MD5).to_lowercase();
```

Comment:
If you implemented the comment above, you can use the hash_reader method to hash the already open file descriptor